### PR TITLE
produce empty tiles when querying outside of area of use of projection

### PR DESCRIPTION
**Pull request by michaelmattig**

### CODE REVIEW

**Comment by michaelmattig** [comment]
[Fixing Projections outside of area of use and general overhaul of GeoTransform and TilingStrategy](https://trello.com/c/6zRZL0k7/361-fixing-projections-outside-of-area-of-use-and-general-overhaul-of-geotransform-and-tilingstrategy)

File: operators/src/source/gdal_source.rs [hunk 1]
```diff
@@ -163,6 +163,56 @@ pub struct GdalDatasetParameters {
...
+/// Default implementation for testing purposes where geo transform doesn't matter
+impl Default for GdalDatasetGeoTransform {
+    fn default() -> Self {
+        Self {
+            origin_coordinate: (0.0, 0.0).into(),
+            x_pixel_size: 1.0,
+            y_pixel_size: -1.0,
+        }
+    }
+}
```
Comment (ChristianBeilschmidt) [inline operators/src/source/gdal_source.rs:186]: I guess we shouldn't implement Default for testing purposes. Maybe we could create a trait TestingDefault. On the other hand, maybe we should use more different values for catching bugs :shrug:.

File: operators/src/adapters/raster_subquery_adapter.rs [hunk 2]
```diff
@@ -198,64 +204,79 @@ where
...
+
+            tile_accu_result
+        } else {
+            // no sub query query rectangle was produced for the current tile, thus output an empty tile
+            Ok(RasterTile2D::new_with_tile_info(
+                this.query_rect.time_interval,
+                fold_tile_spec,
+                GridOrEmpty::Empty(EmptyGrid2D::<PixelType>::new(
+                    fold_tile_spec.tile_size_in_pixels,
+                    this.no_data_value.unwrap_or(PixelType::zero()),
```
Comment (ChristianBeilschmidt) [inline operators/src/adapters/raster_subquery_adapter.rs:276]: maybe add a TODO that this is a fallback assumption

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: datatypes/src/raster/tiling.rs [hunk 3]
```diff
@@ -66,7 +69,9 @@ impl TilingStrategy {
         let lr_idx = self
             .geo_transform
             .coordinate_to_grid_idx_2d(partition.lower_right());
-
+        // TODO: only subtract if lower right coordinate is exactly on pixel edge because
+        //       it is not included in the partition. We don't want to lose the pixel if
+        //       it is actually contained in the partition
```
[the remark this answers is not in the document]
Comment (michaelmattig) [reply datatypes/src/raster/tiling.rs]: 689bfd60e5d22148aac2b05d4f5b6a8e7141092c

File: operators/src/source/gdal_source.rs (hunk 1 above)
Comment (michaelmattig) [reply operators/src/source/gdal_source.rs]: 7a6766483729089fac8b6f0ffcb2c6b1daf168c1

File: operators/src/adapters/raster_subquery_adapter.rs (hunk 2 above)
Comment (michaelmattig) [reply operators/src/adapters/raster_subquery_adapter.rs:277]: 689bfd60e5d22148aac2b05d4f5b6a8e7141092c

**Review by ChristianBeilschmidt — APPROVED** [review]

**Pull request merged**